frontend/src/routes/pipelines/runs.ts

```diff
@@ -0,0 +1,67 @@
+import {
+  routePipelineRunCloneNamespace,
+  routePipelineRunCreateNamespace,
+  routePipelineRunDetailsNamespace,
+  routePipelineRunJobCloneNamespace,
+  routePipelineRunJobDetailsNamespace,
+} from './global';
+import {
+  experimentRunDetailsRoute,
+  experimentScheduleDetailsRoute,
+  experimentsCloneRunRoute,
+  experimentsCloneScheduleRoute,
+  experimentsCreateRunRoute,
+  experimentsScheduleRunRoute,
+} from './experiments';
+
+export const cloneScheduleRoute = (
+  namespace: string,
+  recurringRunId: string,
+  experimentId: string | undefined | false,
+): string =>
+  experimentId
+    ? experimentsCloneScheduleRoute(namespace, experimentId, recurringRunId)
+    : routePipelineRunJobCloneNamespace(namespace, recurringRunId);
+
+export const scheduleRunRoute = (
+  namespace: string | undefined,
+  experimentId: string | undefined | false,
+): string =>
+  experimentId
+    ? experimentsScheduleRunRoute(namespace, experimentId)
+    : routePipelineRunCreateNamespace(namespace);
+
+export const createRunRoute = (
+  namespace: string | undefined,
+  experimentId: string | undefined | false,
+): string =>
+  experimentId
+    ? experimentsCreateRunRoute(namespace, experimentId)
+    : routePipelineRunCreateNamespace(namespace);
+
+export const scheduleDetailsRoute = (
+  namespace: string,
+  recurringRunId: string,
+  experimentId: string | undefined | false,
+): string =>
+  experimentId
+    ? experimentScheduleDetailsRoute(namespace, experimentId, recurringRunId)
+    : routePipelineRunJobDetailsNamespace(namespace, recurringRunId);
+
+export const runDetailsRoute = (
+  namespace: string,
+  runId: string,
+  experimentId: string | undefined | false,
+): string =>
+  experimentId
+    ? experimentRunDetailsRoute(namespace, experimentId, runId)
+    : routePipelineRunDetailsNamespace(namespace, runId);
+
+export const cloneRunRoute = (
+  namespace: string,
+  runId: string,
+  experimentId: string | undefined | false,
```

Comment:
Why is an `id` "false"? and `undefined` should be done with the question mark (but that's more of a nit)

Reply:
Because the usage was `isExperimentAvailable && experimentId`. Now it is `isExperimentAvailable ? experimentId : undefined` so `false` is no longer present. Still not sure why that would be a problem as a possible type.

Reply:
I'm also not sure what you mean by "undefined should be done with the question mark"

Reply:
> Because the usage was isExperimentAvailable && experimentId. Now it is isExperimentAvailable ? experimentId : undefined so false is no longer present. Still not sure why that would be a problem as a possible type.

Do not distort function definitions to meet implementation. Function definitions should always be sane on their own right. It's bad for maintenance for a function to say an `id` can be `false` -- my first question seeing this is "how can an ID be false?"

I get you're saving yourself several characters in several spots by doing this, but you're just adding weird edge-case complexity to the reading of the code to fit your shorter version of using it. We do not pay by the character, so being succinct for loss of clarity is not right imo.

--

> I'm also not sure what you mean by "undefined should be done with the question mark"

Oh I was just saying do this -- which is a nit and not something I care about, just felt odd seeing `| undefined` on a argument to a function.
```
  experimentId?: string,
```

Reply:
Making a parameter optional implies that the function it uses does not require the parameter to work as intended. I disagree that it is odd.

Reply:
I'm not sure I follow your reasoning... if the function cannot run without the param... why are you allowing undefined? Stop before the function is called & make it a required param.

If the function has a failsafe for not having the param and exits early / does nothing, then it clearly does handle undefined.

Optional is undefined. That's not confusing. This is mostly moot, because both are valid ways to say it's "optional". But if you're saying `prop?: string` and `prop: string | undefined` are different, I disagree; syntactic sugar. Just like there is a 100 different ways to code, this is just one of those "preference" things. Which is why it was a nit.

There is a difference in `null` and `undefined` but however you cut it, `undefined` means the same thing used one way or another. "not-explicitly nothing / omitted". `null` means "explicitly nothing"

Reply:
The possibility of `undefined` is due to Params having this option, which these route functions hinge on; https://github.com/remix-run/react-router/blob/9e7486b89e712b765d947297f228650cdc0c488e/packages/router/utils.ts#L373.

Forcing a developer to specify undefined explicitly for a parameter is not the same as not requiring a parameter at all. That's the difference.

Reply:
Hmmm... I think you're touching on something we should fix on the usage side not the implementation side.

Take the usage as:
```javascript
func('foo', undefined);
func('foo', 'bar');
func('foo'); 
```

If func is going to work legit both ways, and do two different things based on the last param. I'd expect it to be defined as:
```javascript
const func = (bar: string, other?: string) => { }
```

If you are forcing the developer to say `undefined` because the function doesn't work at all but is safe guarded. I can see your point in defining it like:
```javascript
const func = (bar: string, other: string | undefined) => { }
```

I get the problem is using `useParams` -- so maybe we need a tech debt item here to fix this up so things like namespace being `| undefined` are not necessary. Maybe we can write a utility to safe guard the undefined nature -- because it legit won't work in this case.
```javascript
export const createRunRoute = (
  namespace: string | undefined,
  experimentId: string | undefined,
): string =>
  experimentId
    ? experimentsCreateRunRoute(namespace, experimentId)
    : routePipelineRunCreateNamespace(namespace);
```
But `experimentId` is wrong imo trying to flex `undefined` as valid value. Since you don't care if it's `undefined` you're judging it on truthyness -- which could be `''` as well. You handle it both ways, so it **is** optional.

Reply:
FWIW, at this point, I think we need a tech debt, this is not a request to change your PR. But I want to be clear on what we are talking about here. @jpuzz0 (cc @Gkrumbach07)

Reply:
I decided to log https://issues.redhat.com/browse/RHOAIENG-4723 to address the concerns Puzzo is bringing up here... let me know if there are any questions.